src/main/java/org/folio/rest/impl/InvoiceHelper.java

```diff
@@ -158,11 +161,44 @@ public Future<Invoice> createInvoice(Invoice invoice, RequestContext requestCont
       })
       .compose(v -> validateAcqUnitsOnCreate(invoice.getAcqUnitIds()))
       .compose(v -> updateWithSystemGeneratedData(invoice))
+      .compose(v -> validateFiscalYearId(invoice, requestContext))
       .compose(v -> invoiceService.createInvoice(invoice, requestContext))
       .map(Invoice::getId)
       .map(invoice::withId);
   }
 
+  private Future<Void> validateFiscalYearId(Invoice invoice, RequestContext requestContext) {
+    if (StringUtils.isNotEmpty(invoice.getFiscalYearId())) {
+      return succeededFuture();
+    }
+
+    Set<String> fundIds = invoice.getAdjustments().stream()
+      .flatMap(adjustment -> adjustment.getFundDistributions().stream())
+      .map(FundDistribution::getFundId)
+      .collect(Collectors.toSet());
+
+    if (!fundIds.isEmpty()) {
+      return HelperUtils.executeWithSemaphores(fundIds,
+          fundId -> currentFiscalYearService.getCurrentFiscalYearByFund(fundId, requestContext),
+          requestContext)
+        .map(fiscalYears -> {
+          List<FiscalYear> uniqueFiscalYears = fiscalYears.stream().distinct().collect(toList());
+          if (uniqueFiscalYears.size() > 1) {
+            String message = String.format(MULTIPLE_FISCAL_YEARS_ERROR.getDescription(), uniqueFiscalYears.get(0).getCode(),
+              uniqueFiscalYears.get(1).getCode());
```

Comment:
It will provide only two values from uniqueFiscalYears list. It would be situations with two or more unique fiscal years

Reply:
Refactored.